src/distances/qgram.jl

```diff
@@ -80,9 +78,141 @@ function _count(s1, s2)
 	return values(d)
 end
 
+# Turn a sequence of qgrams to a count dict for them, i.e. map each
+# qgram to the number of times it has been seen.
+function countdict(qgrams)
+    d = Dict{eltype(qgrams), Int32}()
+    for qg in qgrams
+        index = Base.ht_keyindex2!(d, qg)
+		if index > 0
+			d.age += 1
+			@inbounds d.keys[index] = qg
+			@inbounds d.vals[index] = d.vals[index][1] + 1
+		else
+			@inbounds Base._setindex!(d, 1, qg, -index)
+		end
+    end
+    d
+end
+
+abstract type AbstractQGramCounts{Q,K} end
+q(qc::AbstractQGramCounts{Q,K}) where {Q,K} = Q
+counts(qc::AbstractQGramCounts) = qc.counts
+
+struct QGramDict{Q,K} <: AbstractQGramCounts{Q,K}
+    counts::Dict{K,Int}
+end
+function QGramDict(s::Union{AbstractString, AbstractVector}, q::Integer = 2)
+    @assert q >= 1
+    qgs = qgrams(s, q)
+    QGramDict{q, eltype(qgs)}(countdict(qgs))
+end
+
+# Faster (than QgramDict) with the qgrams presorted
+struct QGramSortedArray{Q,K} <: AbstractQGramCounts{Q,K}
+    counts::Vector{Pair{K,Int}}
+end
+function QGramSortedArray(s::Union{AbstractString, AbstractVector}, q::Integer = 2)
```

Comment:
Could it work with arbitrary iterator that define `length` (like `graphemes` for instance)?

Reply:
I would think so. I'm not very familiar with Unicode or `graphemes` but if your existing functionality supports it I should be able to add it. Will check. Would fulfilling tests like this be what you are after:

````julia
s1, s2 = "alborgów", "amoniak"
p1, p2 = QGramSortedVector(s1), QGramSortedVector(s2)
p1b, p2b = QGramSortedVector(graphemes(s1)), QGramSortedVector(graphemes(s2))
@test evaluate(Jaccard(2), s1, s2) == evaluate(Jaccard(2), p1, p2)
@test evaluate(Jaccard(2), graphemes(s1), graphemes(s2)) == evaluate(Jaccard(2), p1, p2)
@test evaluate(Jaccard(2), graphemes(s1), graphemes(s2)) == evaluate(Jaccard(2), p1b, p2b)
````